security/pkg/k8s/controller/workloadsecret.go

```diff
@@ -359,39 +401,54 @@ func (sc *SecretController) scrtUpdated(oldObj, newObj interface{}) {
 	namespace := scrt.GetNamespace()
 	name := scrt.GetName()
 
-	certBytes := scrt.Data[CertChainID]
-	cert, err := util.ParsePemEncodedCertificate(certBytes)
-	if err != nil {
-		log.Warnf("Failed to parse certificates in secret %s/%s (error: %v), refreshing the secret.",
-			namespace, name, err)
-		if err = sc.refreshSecret(scrt); err != nil {
-			log.Errora(err)
-		}
+	_, waitErr := sc.certUtil.GetWaitTime(scrt.Data[CertChainID], time.Now(), sc.minGracePeriod)
 
-		return
-	}
+	caCert, _, _, rootCertificate := sc.ca.GetCAKeyCertBundle().GetAllPem()
 
-	certLifeTimeLeft := time.Until(cert.NotAfter)
-	certLifeTime := cert.NotAfter.Sub(cert.NotBefore)
-	// TODO(myidpt): we may introduce a minimum gracePeriod, without making the config too complex.
-	// Because time.Duration only takes int type, multiply gracePeriodRatio by 1000 and then divide it.
-	gracePeriod := time.Duration(sc.gracePeriodRatio*1000) * certLifeTime / 1000
-	if gracePeriod < sc.minGracePeriod {
-		log.Warnf("gracePeriod (%v * %f) = %v is less than minGracePeriod %v. Apply minGracePeriod.",
-			certLifeTime, sc.gracePeriodRatio, gracePeriod, sc.minGracePeriod)
-		gracePeriod = sc.minGracePeriod
+	// Check if root certificate in key cert bundle is not up-to-date. With mutiple
+	// Citadel deployed in Istio, the root certificate in istio-ca-secret could be
+	// rotated by any Citadel and become newer than the one in local key cert bundle.
+	if !bytes.Equal(rootCertificate, scrt.Data[RootCertID]) {
+		caSecret, scrtErr := sc.caSecretController.LoadCASecretWithRetry(CASecret,
+			sc.istioCaStorageNamespace, 100*time.Millisecond, 5*time.Second)
+		if scrtErr != nil {
+			log.Errorf("Fail to load CA secret %s:%s (error: %s), skip updating secret %s",
+				sc.istioCaStorageNamespace, CASecret, scrtErr.Error(), name)
+			return
+		}
+		// The CA cert from istio-ca-secret is the source of truth. If CA cert
+		// in local keycertbundle does not match the CA cert in istio-ca-secret,
+		// reload root cert into keycertbundle.
+		if !bytes.Equal(caCert, caSecret.Data[caCertID]) {
+			log.Warn("CA cert in KeyCertBundle does not match CA cert in " +
+				"istio-ca-secret. Start to reload root cert into KeyCertBundle")
+			var err error
+			rootCertificate, err = util.AppendRootCerts(caSecret.Data[caCertID], sc.rootCertFile)
+			if err != nil {
+				log.Errorf("failed to append root certificates: %s", err.Error())
+				return
+			}
```

Comment:
Only the Citadel that rotates the istio-ca-secret needs to read the rootCertFile from the file mount and add it. Other Citadels only need to keep the KCB in sync with the istio-ca-secret.

Reply:
For release-1.1 we don't need to add root cert, because root cert file is only used in plugin cert mode. I have removed appending root cert here and root cert rotator.
For release-1.3, even in self signed mode, the root cert is self signed ca cert appended by root cert from file, and this root cert is not stored in istio-ca-secret. Other Citadels need to sync with istio-ca-secret to get ca cert first, and then append root cert locally to update root cert in their key cert bundle. Maybe we can add root cert into istio-ca-secret.